spices/SPICE-0014-canonical-formatter.adoc

```diff
@@ -0,0 +1,780 @@
+
+== Detailed design
+
+The formatter won't allow any form of customization, similarly to https://pkg.go.dev/cmd/gofmt[gofmt].
+The idea is that there is a single canonical set of rules to format Pkl code. If people want to format their Pkl
+code differently, they can use other tools that allow customization.
+
+The standard library and https://github.com/apple/pkl-pantry[pkl-pantry] will be formatted using `pkl format`.
+
+Syntactically invalid modules cannot be formated. They will return an error when called via the CLI or the Java API.
+A syntactically invalid module is considered to have formatting issues.
+
+The formatter will not simplify code. Many of these simplifications are not possible with only parsing information
+or they may require parse time variable resolution. +
+There's also the problem that some of these simplifications are not wanted by the users. For example: during
+development, a user may add an unused import that they want to require later, but the formatter will remove it.
+
+=== CLI command
+
+The CLI command will receive Pkl files as inputs and either check or format them.
+
+[source,shell]
+----
+pkl format [OPTIONS] [<modules-or-dir>]
+
+  Run the formatter with the given options.
+  Inputs can be a glob of files or a directory which will be recursively searched for Pkl files.
+
+Options:
+  -c, --check        Check if the inputs are properly formatted, printing a diff to stdout.
+                     Returns non-zero in case of failure.
+  -l, --list         Check if the inputs are properly formatted,
+                     printing the file name to stdout in case they are not.
+                     Returns non-zero in case of failure.
+  -r, --replace      Format all the files in place.
+  -f, --format DIR   Format all the files storing the results in the given directory.
+----
+
+=== Java API
+
+The Java API will provide a way to use the formatter programmatically from any JVM language.
+
+[source, java]
+----
+public class PklFormatter {
+  // Check the given file for formatting violations and return
+  // a possible diff of said violations
+  public @Nullable String check(Path path) {}
+  public @Nullable String check(String code) {}
+
+  // Returns a formatted version of the given file
+  public String format(Path path) {}
+  public String format(String code) {}
+}
+----
+
+=== Rules
+
+The formatter will follow a set of rules. These rules are set in stone and cannot be configured.
+It's important to agree on these rules now, because changing them later would cause a big incompatibility problem
+for users, where their whole codebase would be formatted different due to a rule change.
+
+Some of these rules were taken from our https://pkl-lang.org/main/current/style-guide/index.html[style guide] verbatim,
+others had to be made more strict.
+
+Rules are presented here in order of more generic to less generic.
+Less generic rules (bigger numbers) overrule previously defined ones, if they conflict.
```

Comment:
"more generic to less generic" is hard to mentally parse. "general to specific" is probably more clear.